mods/default/nodes.lua

```diff
@@ -1422,115 +1410,200 @@ local function has_locked_chest_privilege(meta, player)
 	return true
 end
 
-minetest.register_node("default:chest", {
-	description = "Chest",
-	tiles = {"default_chest_top.png", "default_chest_top.png", "default_chest_side.png",
-		"default_chest_side.png", "default_chest_side.png", "default_chest_front.png"},
-	paramtype2 = "facedir",
-	groups = {choppy = 2, oddly_breakable_by_hand = 2},
-	legacy_facedir_simple = true,
-	is_ground_content = false,
-	sounds = default.node_sound_wood_defaults(),
+local function chest_lid_obstructed(pos)
+	local above = { x = pos.x, y = pos.y + 1, z = pos.z }
+	local def = minetest.registered_nodes[minetest.get_node(above).name]
+	-- allow ladders, signs, wallmounted things and torches to not obstruct
+	if def.drawtype == "airlike" or
+			def.drawtype == "signlike" or
+			def.drawtype == "torchlike" or
+			(def.drawtype == "nodebox" and def.paramtype2 == "wallmounted") then
+		return false
+	end
+	return true
+end
 
-	on_construct = function(pos)
+local _open = {}
+
+minetest.register_on_player_receive_fields(function(player, formname, fields)
+	if formname ~= "default:chest" then
+		return
+	end
+	if not fields.quit then
+		return
+	end
+	local pn = player:get_player_name()
+
+	local pos = _open[pn].pos
+	local sound = _open[pn].sound
+	local swap = _open[pn].swap
+	local node = minetest.get_node(pos)
+
+	_open[pn] = nil
+	for k, v in pairs(_open) do
+		if v.pos.x == pos.x and v.pos.y == pos.y and v.pos.z == pos.z then
+			return true
+		end
+	end
+	minetest.swap_node(pos, { name = "default:" .. swap, param2 = node.param2 })
+	minetest.sound_play({ name = sound, gain = 0.3, pos = pos, max_hear_distance = 10})
+	return true
+end)
+
+function default.register_chest(name, def)
+	def.drawtype = "mesh"
+	def.visual = "mesh"
+	def.paramtype = "light"
+	def.paramtype2 = "facedir"
+	def.legacy_facedir_simple = true
+	def.is_ground_content = false
+
+	if def.protected then
+		def.on_construct = function(pos)
+			local meta = minetest.get_meta(pos)
+			meta:set_string("infotext", "Locked Chest")
```

Comment:
@Jordach BFG's chest is an entity with animation. This implementation isn't animated on purpose. Hence /clearobjects is irrelevant for this implementation, since the chest always is a mesh node, open or not.

I purposely chose to make a very restrained and simple implementation that uses nodes and not entities, as I believe that adding an object in the game for the purpose of an animation is a bit over the top for minetest_game. The code of this change is pretty much the same as what e.g. the basic doors do, which is fitting, relatively simple and doesn't spawn entities everywhere.

Because we want to use a single bitmap for both open and closed chests, I decided to make a closed and an open mesh and UV unwrap them in the same way, such that chests can use a single texture. This is preferable since people making textures will be able to replace the skin more easily - there is only one and it's easy to understand. This is why there's a second mesh.

Black was deliberately chosen - I personally think it's a good trade off and fitting. Texture pack makers can draw anything they want on the two exposed black surfaced.

Give it a try, though, don't just look at the code and assume it just breaks everywhere.